protocols/gossipsub/src/types.rs

```diff
@@ -71,12 +77,115 @@ impl std::fmt::Debug for MessageId {
+    pub(crate) connections: HashMap<ConnectionId, RpcSender>,
+}
+
+impl PeerConnections {
+    /// Send a `RpcOut::Control` message to the `RpcReceiver`
+    /// this is high priority. By cloning `futures::channel::mpsc::Sender`
+    /// we get one extra slot in the channel's capacity.
+    pub(crate) fn control(&mut self, control: ControlAction) {
+        let mut rpc = RpcOut::Control(control);
+        for sender in self.connections.values_mut() {
+            match sender.priority.clone().try_send(rpc) {
+                Ok(_) => {
+                    return;
+                }
+                Err(err) => {
+                    rpc = err.into_inner();
+                }
+            }
+        }
+    }
+
+    /// Send a `RpcOut::Subscribe` message to the `RpcReceiver`
+    /// this is high priority. By cloning `futures::channel::mpsc::Sender`
+    /// we get one extra slot in the channel's capacity.
+    pub(crate) fn subscribe(&mut self, topic: TopicHash) {
+        let mut rpc = RpcOut::Subscribe(topic);
+        for sender in self.connections.values_mut() {
+            match sender.priority.clone().try_send(rpc) {
+                Ok(_) => {
+                    return;
+                }
+                Err(err) => {
+                    rpc = err.into_inner();
+                }
+            }
+        }
+    }
+
+    /// Send a `RpcOut::Unsubscribe` message to the `RpcReceiver`
+    /// this is high priority. By cloning `futures::channel::mpsc::Sender`
+    /// we get one extra slot in the channel's capacity.
+    pub(crate) fn unsubscribe(&mut self, topic: TopicHash) {
+        let mut rpc = RpcOut::Unsubscribe(topic);
+        for sender in self.connections.values_mut() {
+            match sender.priority.clone().try_send(rpc) {
+                Ok(_) => {
+                    return;
+                }
+                Err(err) => {
+                    rpc = err.into_inner();
+                }
+            }
+        }
+    }
+
+    /// Send a `RpcOut::Publish` message to the `RpcReceiver`
+    /// this is high priority. If message sending fails, an `Err` is returned.
+    pub(crate) fn publish(
+        &mut self,
+        message: RawMessage,
+        metrics: Option<&mut Metrics>,
+    ) -> Result<(), ()> {
+        let mut rpc = RpcOut::Publish(message.clone());
+        for sender in self.connections.values_mut() {
+            match sender.priority.try_send(rpc) {
+                Ok(_) => {
+                    if let Some(m) = metrics {
+                        m.msg_sent(&message.topic, message.raw_protobuf_len());
+                    }
+                    return Ok(());
+                }
+                Err(err) if err.is_full() => return Err(()),
+                // Channel is closed, try another sender.
+                Err(err) => {
+                    rpc = err.into_inner();
+                }
+            }
+        }
+        unreachable!("At least one peer should be available");
```

Comment:
Why is this unreachable? Say that all channels to the given peer are closed, the `for` loop above will not `return` and thus this `unreachable` is hit, no?

Reply:
Say that each connection has an error, e.g. here:

https://github.com/libp2p/rust-libp2p/blob/c3500bb79a57fde067394911bd2cde76c403da9b/swarm/src/connection/pool/task.rs#L237

Then each connection is going to close their channel.

Say that "the same" time (concurrently) the user calls `Behaviour::publish`.

In such case, the `Behaviour` would still track the connections, even though each channel would be dropped.

Does the above make sense?